packages/webcomponentsjs/ts_src/platform/event-listener-options.ts

```diff
@@ -0,0 +1,129 @@
+const supportsEventOptions = (() => {
+  let supported = false;
+  const eventOptions = {
+    get capture() {
+      supported = true;
+      return false;
+    },
+  };
+  const listener = () => {};
+  // NOTE: These will be unpatched at this point.
+  window.addEventListener('test', listener, eventOptions);
+  window.removeEventListener('test', listener, eventOptions);
+  return supported;
+})();
+
+const nativeEventTarget = window.EventTarget ?? window.Node;
+
+if (
+  !supportsEventOptions &&
+  'addEventListener' in nativeEventTarget.prototype
+) {
+  const parseEventOptions = (
+    optionsOrCapture?: boolean | AddEventListenerOptions
+  ) => {
+    let capture, once;
+    if (
+      optionsOrCapture &&
+      (typeof optionsOrCapture === 'object' ||
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (optionsOrCapture as any) instanceof Object)
+    ) {
+      capture = Boolean((optionsOrCapture as AddEventListenerOptions).capture);
+      once = Boolean((optionsOrCapture as AddEventListenerOptions).once);
+    } else {
+      capture = Boolean(optionsOrCapture);
+      once = false;
+    }
+    return {
+      capture,
+      once,
+    };
+  };
+
+  const origAddEventListener = nativeEventTarget.prototype.addEventListener;
+  const origRemoveEventListener =
+    nativeEventTarget.prototype.removeEventListener;
+
+  const captureListenerMap = new WeakMap();
+  const listenerMap = new WeakMap();
+
+  const getListenerMap = (
+    target: EventTarget,
+    type: string,
+    capture: boolean
+  ) => {
+    const elMap = capture ? captureListenerMap : listenerMap;
+    let typeMap = elMap.get(target);
+    if (typeMap === undefined) {
+      elMap.set(target, (typeMap = new Map()));
+    }
+    let listeners = typeMap.get(type);
+    if (listeners === undefined) {
+      typeMap.set(type, (listeners = new WeakMap()));
+    }
+    return listeners;
+  };
+
+  nativeEventTarget.prototype.addEventListener = function (
+    type: string,
+    listener: EventListenerOrEventListenerObject | null,
+    options?: boolean | AddEventListenerOptions | undefined
+  ) {
+    if (listener == null) {
+      return;
+    }
+    const {capture, once} = parseEventOptions(options);
+    const map = getListenerMap(this, type, capture);
+    let wrapper = map.get(listener);
+    if (wrapper === undefined) {
+      wrapper = (e: Event) => {
+        map.delete(listener);
+        // Try to remove both listener and wrapper since `once` state
+        // is not used as a unique key and we don't track which was added.
+        origRemoveEventListener.call(this, type, wrapper, capture);
+        origRemoveEventListener.call(this, type, listener, capture);
+        ((listener as EventListenerObject).handleEvent ?? listener)(e);
+      };
+      map.set(listener, wrapper);
+      const nativeListener = once ? wrapper : listener;
```

Comment:
If you added `nativeListener` to the map instead of `wrapper`, I think you could avoid calling `origRemoveEventListener` twice in remove also, since the correct listener would always be whatever comes out of the map.

Reply:
Done.